### PR TITLE
Add metatransaction broadcast service

### PR DESCRIPTION
**Pull request by area**
This PR introduces a service that can be used to broadcast (and pay for) meta-transactions we want to. Those transactions are:

* Any transaction to a 'core' contract - TokenLocking, ColonyNetwork, any Colony, any Extension
* Any transaction sending or giving control of tokens to one of those contracts.

### Note 

~This PR is pointed at `feat/meta-transactions`. That's to allow this to be reviewed separately from `feat/meat-transactions`. Once #962 is merged, this should be rebased on `develop` and pointed to `develop`. This PR should not be merged anywhere until #962 is merged.~

### CODE REVIEW

File: packages/metatransaction-broadcaster/adapters/discord.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+client.once("ready", async () => {
+  channel = await client.channels.fetch(process.env.DISCORD_CHANNEL_ID);
+});
+
+client.login(process.env.DISCORD_BOT_TOKEN);
+
+const DiscordAdapter = {
+  async log(output) {
+    console.log(output);
+    // channel.send(output);
```
Comment (kronosapiens) [inline packages/metatransaction-broadcaster/adapters/discord.js:14]: Is this intended?

File: packages/metatransaction-broadcaster/package.json [hunk 2]
```diff
@@ -0,0 +1,26 @@
+{
+  "name": "metatransaction-broadcaster",
+  "version": "1.0.0",
+  "description": "",
```
Comment (kronosapiens) [inline packages/metatransaction-broadcaster/package.json:4]: Add description?

File: packages/metatransaction-broadcaster/package.json [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+  "main": "index.js",
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1"
```
Comment (kronosapiens) [inline packages/metatransaction-broadcaster/package.json:7]: Redundant I know but would be nice to fill this in.

File: test/packages/metaTransactionBroadcaster.js [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+      provider,
+    });
+    await broadcaster.initialise(colonyNetwork.address);
+  });
+
+  afterEach(async () => {
+    await broadcaster.close();
+  });
+
+  describe("should correctly identify transactions as valid or not", function () {
```
Comment (kronosapiens) [inline test/packages/metaTransactionBroadcaster.js:61]: Can we show a transaction to a user address is invalid?

File: packages/metatransaction-broadcaster/MetatransactionBroadcaster.js [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+const ConsoleAdapter = require("./adapters/console").default;
+
+class MetatransactionBroadcaster {
+  /**
+   * Constructor for MetatransactionBroadcaster
+   * @param {string} privateKey              The private key of the address that executes the metatransactions
+   * @param {Object} loader                  The loader for loading the contract interfaces. Usually a TruffleLoader.
+   * @param {Object} provider                Ethers provider that allows access to an ethereum node.
+   */
+  constructor({ privateKey, loader, provider, adapter, dbPath = "./mtxCache.sqlite", port = 3000 }) {
```
Comment (kronosapiens) [inline packages/metatransaction-broadcaster/MetatransactionBroadcaster.js:16]: These latter two arguments should be defaults passed in by `index`.

File: packages/metatransaction-broadcaster/MetatransactionBroadcaster.js [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+    this.loader = loader;
+
+    this.adapter = adapter;
+    if (typeof this.adapter === "undefined") {
+      this.adapter = ConsoleAdapter;
+    }
+
+    this.app = express();
+
+    this.app.use(function (req, res, next) {
```
Comment (kronosapiens) [inline packages/metatransaction-broadcaster/MetatransactionBroadcaster.js:33]: Could be neat to put this logic in a common "express utils" folder shared across packages.

File: packages/metatransaction-broadcaster/MetatransactionBroadcaster.js [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+      }
+    });
+
+    this.server = this.app.listen(port, () => {
+      console.log(`⭐️ Metatransaction Broadcaster running on port ${this.server.address().port}`);
+    });
+  }
+
+  /**
+   * Initialises the Kyc Oracle so that it knows where to find the `Whitelist` contract
```
Comment (kronosapiens) [inline packages/metatransaction-broadcaster/MetatransactionBroadcaster.js:97]: Update comment

File: packages/metatransaction-broadcaster/MetatransactionBroadcaster.js [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+  async close() {
+    this.server.close();
+  }
+
+  /**
+   * Update the gas estimate
+   * @param  {string}  Transaction speed (fastest, fast, safeLow)
+   * @return {Promise}
+   */
+  async updateGasEstimate(_type) {
```
Comment (kronosapiens) [inline packages/metatransaction-broadcaster/MetatransactionBroadcaster.js:125]: Another good function to put in "express utils"

File: packages/metatransaction-broadcaster/adapters/discord.js (hunk 1 above)
Comment (area) [reply packages/metatransaction-broadcaster/adapters/discord.js:14]: Nope :(

File: packages/metatransaction-broadcaster/package.json (hunk 3 above)
Comment (area) [reply packages/metatransaction-broadcaster/package.json:7]: With what, exactly?

Comment (kronosapiens) [reply packages/metatransaction-broadcaster/package.json:7]: Nothing, there's no other test commands.

File: test/packages/metaTransactionBroadcaster.js [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+
+      expect(valid).to.be.equal(false);
+
+      // Going to a colony is okay though
+      txData = await metaTxToken.contract.methods.transfer(colony.address, 300000).encodeABI();
+      valid = await broadcaster.isTransactionValid(metaTxToken.address, txData);
+
+      expect(valid).to.be.equal(true);
+    });
+    it("transactions to a token are accepted base on destination address for transferFrom", async function () {
```
Comment (kronosapiens) [inline test/packages/metaTransactionBroadcaster.js:125]: `based`

**Review by kronosapiens — COMMENTED** [review]
Very nice. I especially like the introduction of `package-utils` here.

File: test/packages/metaTransactionBroadcaster.js [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+
+      const tokenLockingAddress = await colonyNetwork.getTokenLocking();
+      valid = await broadcaster.isAddressValid(tokenLockingAddress);
+      expect(valid).to.be.equal(true);
+    });
+
+    it("transactions to a colony are accepted", async function () {
+      let valid = await broadcaster.isAddressValid(colony.address);
+      expect(valid).to.be.equal(true);
+      valid = await broadcaster.isAddressValid(colony.address);
```
[the remark this answers is not in the document]
Comment (area) [reply test/packages/metaTransactionBroadcaster.js:74]: The first time to be worked out from scratch, the second time to hit the cache.

File: test/packages/metaTransactionBroadcaster.js [hunk 11]
```diff
@@ -0,0 +1,240 @@
...
+  const USER1 = accounts[1];
+
+  let colonyNetwork;
+  let colony;
+
+  let broadcaster;
+  let metaTxToken;
+
+  beforeEach(async () => {
+    colonyNetwork = await setupColonyNetwork();
```
[the remark this answers is not in the document]
Comment (area) [reply test/packages/metaTransactionBroadcaster.js:45]: I've improved this, but snapshotting only applies between test files, not between tests, so some amount of this is still required.

**Review by kronosapiens — DISMISSED** [review]

**Comment by area** [comment]
I think we need to inspect metatransactions for ones that execute `makeArbitraryTransaction` to prevent people setting up a colony for themselves, and using it as a wallet that we pay gas fees for. I'm struggling to come up with a solution though, and might just have to say we don't pay for `makeArbitraryTransaction`. The scenario I can't find a clean solution to is making an arbitrary transaction to another colony, which executes another contract (possibly another colony, which executes...)

**Comment by kronosapiens** [comment]
That's fair. I think we can get away with saying that "arbitrary transactions are an advanced feature that we cannot pay for, for security reasons", which is the truth.

**Review by kronosapiens — APPROVED** [review]

File: test/packages/metaTransactionBroadcaster.js [hunk 12]
```diff
@@ -0,0 +1,388 @@
...
+
+      expect(ethBalanceAfter).to.eq.BN(ethBalanceBefore);
+    });
+
+    it("a transaction that would be valid but is too expensive is rejected and not mined", async function () {
+      const extensionImplementation = await GasGuzzler.new();
+      const resolver = await Resolver.new();
+      await setupEtherRouter("GasGuzzler", { GasGuzzler: extensionImplementation.address }, resolver);
+      const TEST_EXTENSION = soliditySha3("GasGuzzler");
+      // await colonyNetwork.createMetaColony(metaTxToken.address);
```
Comment (kronosapiens) [inline test/packages/metaTransactionBroadcaster.js:336]: Oops

File: test/packages/metaTransactionBroadcaster.js [hunk 13]
```diff
@@ -0,0 +1,388 @@
...
+
+      const mcAddress = await colonyNetwork.getMetaColony();
+      const metaColony = await IMetaColony.at(mcAddress);
+
+      await metaColony.addExtensionToNetwork(TEST_EXTENSION, resolver.address);
+
+      await colony.installExtension(TEST_EXTENSION, 1);
+
+      const extensionAddress = await colonyNetwork.getExtensionInstallation(TEST_EXTENSION, colony.address);
+      // console.log(extensionAddress)
```
Comment (kronosapiens) [inline test/packages/metaTransactionBroadcaster.js:346]: Ditto

**Review by kronosapiens — APPROVED** [review]
Nicely done

File: packages/metatransaction-broadcaster/MetatransactionBroadcaster.js [hunk 14]
```diff
@@ -0,0 +1,266 @@
...
+          const data = {};
+          if (!addressValid) {
+            data.target = "Not a contract we pay metatransactions for";
+          }
+          if (!allowedColonyFamilyTransaction) {
+            data.payload = "Not a function on colony we pay metatransactions for";
+          }
+
+          // validTokenTransaction after '||' unnecessary, but included for clarity
+          // This condition guards against a colony being set up with a native token that follows
```
[the remark this answers is not in the document]
Comment (area) [reply packages/metatransaction-broadcaster/MetatransactionBroadcaster.js:83]: I think we're in agreement here and there's just a mismatch of language, but we can talk about it.

File: packages/metatransaction-broadcaster/MetatransactionBroadcaster.js [hunk 15]
```diff
@@ -0,0 +1,266 @@
...
+      }
+    } catch (err) {
+      // Not an extension
+    }
+
+    // TODO Is it a new-style extension?
+    return false;
+  }
+
+  async setAddressValid(address) {
```
[the remark this answers is not in the document]
Comment (area) [reply packages/metatransaction-broadcaster/MetatransactionBroadcaster.js:204]: Hah, fair, I guess where this has ended up it's not needed. I think the intention was that the table would hold both true and false, and then either making an address valid or invalid would both be an `INSERT` command. I'm inclined to leave it as it is though.

**Review by kronosapiens — APPROVED** [review]

**Pull request merged**